### PR TITLE
Add statuscode check for 200/404 pages

### PR DESCRIPTION
**Pull request by barryvdh**
Check the statuscode of the current page + a page that doesn't exist.

It would perhaps be better if the Response object was also available in the Rule, so the statuscode and headers can be checked.

### CODE REVIEW

File: app/Rules/StatusCode404.php [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class StatusCode404 extends Rule
+{
+    protected $statusCode;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function check()
+    {
+        $uri = rtrim($this->url, '/').'/page-to-test-404-response';
```
Comment (m1guelpf) [inline app/Rules/StatusCode404.php]: Should that be hardcoded?

Comment (barryvdh) [inline app/Rules/StatusCode404.php]: What would you suggest? We just need a route that doesn't usually exist. So we could make it more unlikely (`/__page-to-test-404-response-for-kupo`) or actually random (`'/page-to-test-404-response-for-kupo-'Str::random(12)`), but I wanted to make it clear in the logs that this isn't an actual 404 from a visitor. And random pages would clutter the logs more probably.

Comment (barryvdh) [inline app/Rules/StatusCode404.php]: I made it a bit more unique; `/-~!page-to-test-404-responses-for-invalid-pages!~-`
But not that you don't have to do anything on your server for this, it should work as long as the page doesn't actually exist.

Comment (m1guelpf) [inline app/Rules/StatusCode404.php]: I mean, you could test if an URL is passed and if it isn't, check for the ramdom one.

**Comment by barryvdh** [comment]
We're already checking the regular URL from the form input for 200 status, so why check another one?

**Comment by m1guelpf** [comment]
@barryvdh Well, what is the 404 check for then?
I found it useful for extending it and making complex checks line #15

**Comment by barryvdh** [comment]
You would usually mostly check the landing page or some important page, you don't usually actively check non existent pages. So this checks it for you, by checking a very unlikely URL, nested upon the original url.

If you explain your use case, perhaps I'll understand why you need that non-existent path configurable.

Checking batches does the opposite, eg. make sure it's valid (see the 200 check)

**Comment by m1guelpf** [comment]
@barryvdh I still don't understand the need to check if a particular page doesn't exist...

**Comment by phanan** [comment]
This is a great idea. Can we somehow add a test, though?

**Comment by barryvdh** [comment]
I can add one for the 200 check, but not sure on the 404. I'd like to inject the UrlFetcher, but the interface doesn't really allow it.

**Comment by barryvdh** [comment]
And the 200 check might be a bit unneeded, as the crawler will fail on 404 pages. So either we allow 404 pages etc to be checked (for the rest of the content), or we can remove the 200 check.

**Comment by barryvdh** [comment]
Removed the 200 check, renamed to 404 to PageNotFoundGives404

**Comment by barryvdh** [comment]
Added tests and uses DI injection for the client now :)

**Comment by phanan** [comment]
Thanks!

**Pull request merged**